File: exercise_02/my3DDatatypes/Vec3D.java
package my3DDatatypes;

public class Vec3D {
	float x, y, z, w = 1;

	public Vec3D(float xx, float yy, float zz) {
		x = xx;
		y = yy;
		z = zz;
	}

	public Vec3D(float xx, float yy, float zz, float ww) {
		this(xx, yy, zz);
		w = ww;
	}

	public Vec3D add(Vec3D v) {
		return new Vec3D(x + v.x, y + v.y, z + v.z);
	}

	public Vec3D minus(Vec3D v) {
		return new Vec3D(x - v.x, y - v.y, z - v.z);
	}

	public Vec3D mult(float a) {
		return new Vec3D(a * x, a * y, a * z);
	}

	public void normalize() {
		float dist = (float) Math.sqrt((x * x) + (y * y) + (z * z));
		x = x / dist;
		y = y / dist;
		z = z / dist;
	}

	public float dot(Vec3D v) {
		return x * v.x + y * v.y + z * v.z;
	}

	public Vec3D cross(Vec3D v) {
		return new Vec3D(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
	}
}

File: exercise_02_cpp/Vec3D.cpp
#include <cmath>
#include "Vec3D.h"

namespace my3DDatatypes {

	Vec3D::Vec3D() { }

	Vec3D::Vec3D(float xx, float yy, float zz) {
		x = xx;
		y = yy;
		z = zz;
	}

	Vec3D::Vec3D(float xx, float yy, float zz, float ww) {
		// major c++ bug Vec3D(xx, yy, zz);
		x = xx;
		y = yy;
		z = zz;
		w = ww;
	}

	Vec3D Vec3D::add(const Vec3D& v) {
		return Vec3D(x + v.x, y + v.y, z + v.z);
	}

	Vec3D Vec3D::minus(const Vec3D& v) {
		return Vec3D(x - v.x, y - v.y, z - v.z);
	}

	Vec3D Vec3D::mult(float a) {
		return Vec3D(a * x, a * y, a * z);
	}

	void Vec3D::normalize() {
		float dist = (float) std::sqrt((x * x) + (y * y) + (z * z));
		x = x / dist;
		y = y / dist;
		z = z / dist;
	}

	float Vec3D::dot(const Vec3D& v) {
		return x * v.x + y * v.y + z * v.z;
	}

	Vec3D Vec3D::cross(const Vec3D& v) {
		return Vec3D(y * v.z - z * v.y, z * v.x - x * v.z, x * v.y - y * v.x);
	}
}
